keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,934 @@
+
+Once the node has transitioned to the NotReady state, it will be reverted to the ReadyState once when the node's capacity is reconfigured to match or exceed the last valid configuration.
+In this case, valid configuration refers to a state which can either be previous hot-plug capacity or the initial capacity in case there was no history of hotplug.
+
+#### Flow Control
+
+```
+T=0: Node initial Resources:
+    - Memory: 10G
+    - Pod: Memory
+
+T=1: Resize Instance to Hotplug Memory
+    - Current Memory: 10G
+    - Update Memory: 15G
+    - Node state: Ready
+
+T=2: Resize Instance to HotUnplug Memory
+    - Current Memory: 15G
+    - UpdatedMemory: 5G
+    - Node state: NotReady
+
+T=3: Resize Instance to Hotplug Memory
+    - Current Memory: 5G
+    - Updated Memory Size: 15G
+    - Node state: Ready
+```
+
+Few of the concerns surrounding hotunplug are listed below
+* Pod re-admission:
+    * Given that there is probability that the current Pod resource usage may exceed the available capacity of node, its necessary to check if the pod can continue Running
+      or if it has to be terminated due to resource crunch.
+* Recalculate OOM adjust score and Swap limits:
+    * Since the total capacity of the node has changed, values associated with the nodes memory capacity must be recomputed.
+* Handling unplug of reserved CPUs.
+
+we intend to propose a separate KEP dedicated to hotunplug of resources to address the same.
+
+**Proposed Code changes**
+
+**Pseudocode for Resource Hotplug**
+
+```go
+func (kl *Kubelet) syncLoopIteration(ctx context.Context, configCh <-chan kubetypes.PodUpdate, handler SyncHandler,
+syncCh <-chan time.Time, housekeepingCh <-chan time.Time, plegCh <-chan *pleg.PodLifecycleEvent) bool {
+    .
+    .
+    case machineInfo := <-kl.nodeResourceManager.MachineInfo():
+        // Resize the containers.
+        klog.InfoS("Resizing containers due to change in MachineInfo")
+        if err := resizeContainers(); err != nil {
+            klog.ErrorS(err, "Failed to resize containers with change in machine info")
+            kl.recorder.Eventf(kl.nodeRef, v1.EventTypeWarning, events.FailedNodeResize, err.Error())
+            break
+        }
+        
+        // Resync the resource managers.
+        klog.InfoS("ResyncComponents resource managers because of change in MachineInfo")
+        if err := kl.containerManager.ResyncComponents(machineInfo); err != nil {
+            klog.ErrorS(err, "Failed to resync resource managers with machine info update")
+            kl.recorder.Eventf(kl.nodeRef, v1.EventTypeWarning, events.FailedNodeResize, err.Error())
+            break
+        }
+        
+        // Update the cached MachineInfo.
+        kl.setCachedMachineInfo(machineInfo)
+    .
+    .
+}
+```
+
+**Changes to resource managers to adapt to hotplug of resources**
```

Comment:
Do we foresee to reuse this function for resource hotunplug in the future? (I'd expect yes)

Reply:
Yes, That's the idea. Apart from code that specifically deals with hot plug and hot-unplug we intend to reuse bits that are applicable for both the scenarios.